### PR TITLE
Topic/default executor

### PR DESCRIPTION
**Pull request by bryce-anderson**
closes #569
This PR changes the mode of management of the client `ExecutorService`. Previously we would make our own "special" executor that was tagged with a trait so that we could tell if we made it and then shut it down. After this PR we pass an optional `customExecutor` in and if its `None`, make one and take responsibility for shutting it down.

### CODE REVIEW

**Comment by bryce-anderson** [comment]
I think this could go a little further and apply the same strategy to the servers and in the same pass boot the scalaz default strategy executor from our code base: that thing is ridiculous with its very small thread count and propensity to deadlock.

File: blaze-client/src/main/scala/org/http4s/client/blaze/package.scala [hunk 1]
```diff
@@ -7,6 +7,7 @@ package object blaze {
 
   /** Default blaze client
     *
-    * This client will create a new connection for every request. */
-  lazy val defaultClient: Client = SimpleHttp1Client(BlazeClientConfig.defaultConfig())
+    * This client will create a new connection for every request.
+    */
+  lazy val defaultClient: Client = SimpleHttp1Client(BlazeClientConfig.defaultConfig)
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/client/blaze/package.scala]: Not apropros to this ticket, but I think we should deprecate this.

File: blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala [hunk 2]
```diff
@@ -13,18 +15,11 @@ import scalaz.{-\/, \/-}
...
 
-  def apply[A <: BlazeConnection](manager: ConnectionManager[A], config: BlazeClientConfig): Client = {
-
-    val shutdownTask = manager.shutdown().flatMap(_ => Task.delay {
-      // shutdown executor services that have been implicitly created for us
-      config.executor match {
-        case es: DefaultExecutorService =>
-          logger.info(s"Shutting down default ExecutorService: $es")
-          es.shutdown()
+  def apply[A <: BlazeConnection](manager: ConnectionManager[A],
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala]: This is shutting down a connection manager it didn't create, and an opaque task that it didn't create.  That violates the "you opened it, you close it" mantra.  Should this be a private interface?

**Comment by rossabaker** [comment]
I think the general direction is correct here.  And I agree that a similar issue exists on the server.

File: blaze-client/src/main/scala/org/http4s/client/blaze/package.scala (hunk 1 above)
Comment (bryce-anderson) [inline blaze-client/src/main/scala/org/http4s/client/blaze/package.scala]: Agreed. In fact, I had deprecated it but decided that it should be a separate PR so I reverted that change. :)

File: blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala (hunk 2 above)
Comment (bryce-anderson) [inline blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala]: Thats a good point about the manager. I'm happy to leave it open if we require the manager shutdown to be a part of the shutdown Task but also open to making it private. I left it open for the off chance that someone wants to make a custom `ConnectionManager` but that seems pretty uncommon so I can go either way.

**Comment by bryce-anderson** [comment]
Do we want to try and address the server in this PR as well or make it a separate one?

**Comment by rossabaker** [comment]
PRs are cheap, progress is not. :+1: and follow up with another.
